staging/src/k8s.io/api/discovery/v1beta1/well_known_labels.go

```diff
@@ -25,4 +25,8 @@ const (
 	// same cluster. It is highly recommended to configure this label for all
 	// EndpointSlices.
 	LabelManagedBy = "endpointslice.kubernetes.io/managed-by"
+	// LabelSkipMirror can be set to true on an Endpoints resource to indicate
+	// that the EndpointSliceMirroring controller should not mirror this
+	// resource with EndpointSlices.
+	LabelSkipMirror = "endpointslice.kubernetes.io/skip-mirror"
```

Comment:
I am wondering if the controller can look up the corresponding service in the cache to see if it has a selector or not, and decide whether it needs to mirror the endpoints. Instead of relying on the label. I guess these are the following scenarios.

1. Service with selector. => no need to mirror
2. Service without selector AND no Endpoints => no need to mirror
3. Service without selector AND has an custom Endpoints object AND no skip mirror label => need to mirror
4. Service without selector AND has an custom Endpoints object AND has skip mirror label => no need to mirror

Reply:
This would be a relatively significant deviation from the KEP, but I think it could be worthwhile. This would mean I could remove all changes from the Endpoints controller as part of this PR, and maybe more significantly, would remove the need for the significant startup wait sequence for this controller. If there were any bugs in that startup wait sequence the controller may never actually start up. It looks like @thockin may already [suggested something like this here](https://github.com/kubernetes/enhancements/pull/1713#discussion_r426089134). If there are no objections, I'll start moving towards this approach.

Reply:
I've updated this PR to use the suggested approach above, I think it is better than requiring the skip-mirror label to be set everywhere and blocking controller startup for that. One place I have not changed that is the self-referential API Server EndpointSlice. The `kubernetes` EndpointSlice is essentially a mirror of the `kubernetes` Endpoints resource, but it's managed by `pkg/master` and not the EndpointSliceMirroring controller. Changing that to use this controller would be messy at best. It also feels backwards to have API Server be dependent on kube-controller-manager.

Reply:
mostly curiosity, is there a case of endpoints without services associated?

Reply:
Great question. I can't think of any that are officially supported, though of course it can happen. From what I can tell, kube-proxy will disregard any endpoints that don't have an associated service. (It loops through services and gets any associated endpoints).

Reply:
just for clarification, the question was unrelated to this PR :smile:

Reply:
Endpoints without a same-name service are automatically removed eventually